sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsPartitions.java

```
@@ -0,0 +1,96 @@
+ * A partition is composed of identifier and properties,
+ * and properties contains metadata information of the partition.
+ * <p>
+ * These APIs are used to modify table partition identifier or partition metadata,
+ * in some cases, they will change actual value of table data as well.
+ *
+ * @since 3.0.0
+ */
+@Experimental
+public interface SupportsPartitions extends Table {
+
+    /**
+     * Create a partition in table.
+     *
+     * @param ident a new partition identifier
+     * @param properties the metadata of a partition
+     * @throws PartitionAlreadyExistsException If a partition already exists for the identifier
+     */
+    void createPartition(
+            InternalRow ident,
+            Map<String, String> properties) throws PartitionAlreadyExistsException;
+
+    /**
+     * Drop a partition from table.
+     *
+     * @param ident a partition identifier
+     * @return true if a partition was deleted, false if no partition exists for the identifier
+     */
+    Boolean dropPartition(InternalRow ident);
+
+    /**
+     * Rename a Partition from old identifier to new identifier with no metadata changed.
+     *
+     * @param oldIdent the partition identifier of the existing partition
+     * @param newIdent the new partition identifier of the partition
+     * @throws NoSuchPartitionException If the partition identifier to rename doesn't exist
+     * @throws PartitionAlreadyExistsException If the new partition identifier already exists
+     */
+    void renamePartition(
+            InternalRow oldIdent,
+            InternalRow newIdent) throws NoSuchPartitionException, PartitionAlreadyExistsException;
+
+    /**
+     * Replace the partition metadata of the existing partition.
+     *
+     * @param ident the partition identifier of the existing partition
+     * @param properties the new metadata of the partition
+     * @throws NoSuchPartitionException If the partition identifier to rename doesn't exist
```

Should this document that implementations can throw `UnsupportedOperationException` if metadata is not supported?

---

yeap.

---

Looks good.